src/managers/auth.py

```diff
@@ -25,85 +27,148 @@ class Acl:
         return {
             "username": self.username,
             "operation": self.operation,
-            "resource": self.resource_str,
+            "resource": self.resource,
         }
 
 
+@dataclass
+class UserCredentials:
+    """Object for representing credentials of a Karapace user."""
+
+    username: str
+    algorithm: Algorithm
+    salt: str
+    password_hash: str
+
+    @property
+    def creds_repr(self) -> dict[str, str]:
+        """Returns the dict with an credentials representation."""
+        return {
+            "username": self.username,
+            "algorithm": self.algorithm,
+            "salt": self.salt,
+            "password_hash": self.password_hash,
+        }
+
+
+class AuthDictEntry(TypedDict):
+    """Data entry for internal auth state on KarapaceAuth."""
+
+    credentials: UserCredentials
+    acls: list[Acl]
+
+
 class KarapaceAuth:
-    """Object for updating Karapace users and ACLs."""
+    """Object for updating Karapace users and ACLs.
+
+    The class will load authfile.json to create an internal mapping, after this, operations are
+    only done internally. To persist changes, `write_authfile` has to be invoked e.g.:
+
+    ```
+        auth = KarapaceAuth(context, workload)
+        auth.add_user(username, password)
+        auth.add_acl()
+
+        auth.write_authfile()
+    ```
+    """
 
     def __init__(self, context: ClusterContext, workload: WorkloadBase):
         self.context = context
         self.workload = workload
 
-    @property
-    def parsed_authfile(self) -> dict:
-        """Return authfile parsed as a dict."""
-        return json.loads("\n".join(self.workload.read(self.workload.paths.registry_authfile)))
+        # Load current users and ACLs to have an internal mapping of auth state.
+        self._load_authfile()
 
-    @property
-    def users(self) -> list[str]:
-        """Return existing users on karapace authfile."""
-        full_users = self.parsed_authfile.get("users", [])
-        return [user["username"] for user in full_users]
+    def _load_authfile(self) -> None:
+        """Load current Karapace authfile."""
+        # Internal state of auth to the class
+        self.auth_dict: dict[str, AuthDictEntry] = {}
 
-    @property
-    def permissions(self) -> list[dict[str, str]] | None:
-        """Return existing permissions on karapace authfile."""
-        return self.parsed_authfile.get("permissions")
+        raw_file = self.workload.read(self.workload.paths.registry_authfile)
+        if not raw_file:
+            return
+        authfile = json.loads("\n".join(raw_file))
+
+        users = [UserCredentials(**user) for user in authfile["users"]]
+        acls = [Acl(**acl) for acl in authfile["permissions"]]
+
+        for user in users:
+            self.auth_dict[user.username] = AuthDictEntry(
+                credentials=user,
+                acls=[acl for acl in acls if acl.username == user.username],
+            )
```

Comment:
**thought:** (non-blocking) I get the feeling that this structure is overly complicated. I don't have a better idea off the top of my head so it's definitely fine to keep, but I wonder if the 'gain' of being able to reference `auth_dict[username]` is worth the JSON/Dict transformations.

Reply:
The problem addressed here is that we don't have a way to link a single username to both credentials and acls with the original json format. If I don't merge them into the internal dictionary here, then I would need to always iterate over all acls to find the ones that have `username == <my-username>`.

With the current approach, it's only done once when loading the file.